File: recursion/strassen/py/test_strassen.py
import strassen
import unittest


class StrassenTest(unittest.TestCase):
    def testNextPowerOfTwo(self):
        self.assertEqual(strassen.next_power_of_two(0), 1)
        self.assertEqual(strassen.next_power_of_two(1), 1)
        self.assertEqual(strassen.next_power_of_two(2), 2)
        self.assertEqual(strassen.next_power_of_two(3), 4)
        self.assertEqual(strassen.next_power_of_two(13), 16)

    def testAdd(self):
        A = [[1,2],[3,4]]
        B = [[1,1],[1,1]]
        expected = [[2,3],[4,5]]
        self.assertEqual(strassen.add(A,B), expected)

    def testMinus(self):
        A = [[1,2],[3,4]]
        B = [[1,1],[1,1]]
        expected = [[0,1],[2,3]]
        self.assertEqual(strassen.minus(A,B), expected)

    def testCopyMatrix(self):
        A = [[1,1],[2,2]]
        B = strassen.init_matrix(4)
        strassen.copy_matrix(A, B, len(A), 0, 0)
        expected = [[1,1,0,0],
                    [2,2,0,0],
                    [0,0,0,0],
                    [0,0,0,0]]
        self.assertEqual(B, expected)
        C = strassen.init_matrix(4)
        strassen.copy_matrix(A, C, len(A), 1, 1)
        expected = [[0,0,0,0],
                    [0,1,1,0],
                    [0,2,2,0],
                    [0,0,0,0]]
        self.assertEqual(C, expected)

    def testGetBlockMatrix(self):
        A = [[1,2],[3,4]]
        a,b,c,d = strassen.get_block_matrix(A)
        self.assertEqual(a, [[1]])
        self.assertEqual(b, [[2]])
        self.assertEqual(c, [[3]])
        self.assertEqual(d, [[4]])

    def testStrassenEven(self):
        A = [[1,1],[1,1]]
        B = [[1,2],[3,4]]
        expected = [[4,6],[4,6]]
        self.assertEqual(strassen.strassen(A,B), expected)

    def testStrassenOdd(self):
        A = [[1,2,3],[2,3,4],[3,4,5]]
        B = [[1,1,1],[1,1,1],[1,1,1]]
        expected = [[6,6,6],[9,9,9],[12,12,12]]
        self.assertEqual(strassen.strassen(A,B), expected)


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: recursion/strassen/java/TestStrassen.java
import org.junit.*;
import static org.junit.Assert.*;
import static org.junit.matchers.JUnitMatchers.*;

import org.javatuples.*;

public class TestStrassen {
  @Test
  public void NextPowerOfTwo() {
    assertEquals(1, Strassen.nextPowerOfTwo(0));
    assertEquals(1, Strassen.nextPowerOfTwo(1));
    assertEquals(2, Strassen.nextPowerOfTwo(2));
    assertEquals(4, Strassen.nextPowerOfTwo(3));
    assertEquals(16, Strassen.nextPowerOfTwo(13));
    assertEquals(16, Strassen.nextPowerOfTwo(16));
  }

  @Test
  public void Add() {
    int[][] A = {{1,2},{3,4}};
    int[][] B = {{1,1},{1,1}};
    int[][] expected = {{2,3},{4,5}};
    assertArrayEquals(expected, Strassen.add(A,B));
  }

  @Test
  public void Minus() {
    int[][] A = {{1,2},{3,4}};
    int[][] B = {{1,1},{1,1}};
    int[][] expected = {{0,1},{2,3}};
    assertArrayEquals(expected, Strassen.minus(A,B));
  }

  @Test
  public void CopyMatrix() {
    int[][] A = {{1,1},{2,2}};
    int[][] B = new int[4][4];
    Strassen.copyMatrix(A, B, A.length, 0, 0);
    int[][] expected_b = new int[][]{{1,1,0,0},
                                     {2,2,0,0},
                                     {0,0,0,0},
                                     {0,0,0,0}};
    assertArrayEquals(expected_b, B);
    int[][] C = new int[4][4];
    Strassen.copyMatrix(A, C, A.length, 1, 1);
    int[][] expected_c = new int[][]{{0,0,0,0},
                                     {0,1,1,0},
                                     {0,2,2,0},
                                     {0,0,0,0}};
    assertArrayEquals(expected_c, C);
  }

  @Test
  public void GetBlockMatrix() {
    int[][] A = {{1,2},{3,4}};
    Quartet<int[][],int[][],int[][],int[][]> all = Strassen.getBlockMatrix(A);
    assertArrayEquals(new int[][]{{1}}, all.getValue0());
    assertArrayEquals(new int[][]{{2}}, all.getValue1());
    assertArrayEquals(new int[][]{{3}}, all.getValue2());
    assertArrayEquals(new int[][]{{4}}, all.getValue3());
  }

  @Test
  public void StrassenEven() {
    int[][] A = {{1,1},{1,1}};
    int[][] B = {{1,2},{3,4}};
    int[][] expected = {{4,6},{4,6}};
    assertArrayEquals(expected, Strassen.strassen(A,B));
  }

  @Test
  public void StrassenOdd() {
    int[][] A = {{1,2,3},{2,3,4},{3,4,5}};
    int[][] B = {{1,1,1},{1,1,1},{1,1,1}};
    int[][] expected = {{6,6,6},{9,9,9},{12,12,12}};
    assertArrayEquals(expected, Strassen.strassen(A,B));
  }

  public static void main(String args[]) {
    org.junit.runner.JUnitCore.main("TestStrassen");
  }
}
